Fix the lack of direct GPU to GPU communications in multi-device runs.

therault opened the pull request:
In https://github.com/ICLDisco/parsec/pull/570, we moved from using cuda_index to using device_index in the 'nvlink' mask that decides if we can directly communicate to another GPU. However, this bitmask was initialized at query time, before devices get assigned a device_index. As a consequence, the bitmask was wrong and no direct device to device communication was happening.

In this PR, we add a step, after all devices have been registered, to complete this initialization.

An alternative is to come back to using cuda_index-based bitmasks, but then the decision if two GPUs can directly communicate is device-type specific and needs to be moved from device_gpu.c to device_cuda_module.c, which means we add another function call to device_cuda_module.c inside the stage_in() function of device_gpu.c

bosilca commented:
Why did we moved from using cuda_index to using the device_index ? We cannot use d2d on devices of different types, and we check for that in `parsec_default_gpu_stage_in` and `parsec_default_gpu_stage_out`, so using the device_index in the mask makes little sense. Going back and rebuilding this mask based on cuda_index seems like a simpler and most resilient solution (we could change the device_index without having to rebuild all the info).

therault commented:
device_gpu.c doesn't know the cuda_index (or hip_index or
level_zero_index), that's the reason why we moved that mask to global
device_index based.

I don't think we want to replicate the full parsec_device_gpu_stage_in()
function in all device implementations, so if we want to go back to
cuda_index based bitmask, we can replace the test

`if( gpu_device->peer_access_mask & (1 <<
candidate_dev->super.device_index) )`
(and similar tests in device_gpu.c)
by a call to the device-specific implementation, something like
`if( gpu_device->peer_access(candidate_dev)`

I'm fine with that, but I thought the global device_index mask would save
these function calls that cannot be inlined in stage_in, which is taking a
lot of time already.



On Fri, Mar 8, 2024 at 3:11 PM bosilca ***@***.***> wrote:

> Why did we moved from using cuda_index to using the device_index ? We
> cannot use d2d on devices of different types, and we check for that in
> parsec_default_gpu_stage_in and parsec_default_gpu_stage_out, so using
> the device_index in the mask makes little sense. Going back and rebuilding
> this mask based on cuda_index seems like a simpler and most resilient
> solution (we could change the device_index without having to rebuild all
> the info).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ICLDisco/parsec/pull/642#issuecomment-1986352426>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABFEZNVU76DEPMU6TOVWMODYXILN5AVCNFSM6AAAAABENLG62WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBWGM2TENBSGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

abouteiller commented:
Passes all ctests and works with large size POTRF -g 8 with nvlink active on Leconte.

devreal commented on parsec/mca/device/device_gpu.c:
```diff
@@ -1510,11 +1524,14 @@ parsec_device_data_stage_in( parsec_device_gpu_module_t* gpu_device,
     /* update the data version in GPU immediately, and mark the data under transfer */
     assert((gpu_elem->version != candidate->version) || (gpu_elem->data_transfer_status == PARSEC_DATA_STATUS_NOT_TRANSFER));
-    gpu_elem->version = candidate->version;
+    if( PARSEC_FLOW_ACCESS_WRITE & type && gpu_task->task_type != PARSEC_GPU_TASK_TYPE_PREFETCH )
+        gpu_elem->version = candidate->version + 1;
+    else
+        gpu_elem->version = candidate->version;
```
Can you please document why we set the version the way we set it? This is delicate and not obvious.

devreal commented on parsec/mca/device/device_gpu.c:
```diff
@@ -1510,11 +1524,14 @@ parsec_device_data_stage_in( parsec_device_gpu_module_t* gpu_device,
     /* update the data version in GPU immediately, and mark the data under transfer */
     assert((gpu_elem->version != candidate->version) || (gpu_elem->data_transfer_status == PARSEC_DATA_STATUS_NOT_TRANSFER));
-    gpu_elem->version = candidate->version;
+    if( PARSEC_FLOW_ACCESS_WRITE & type && gpu_task->task_type != PARSEC_GPU_TASK_TYPE_PREFETCH )
```
For readability

```suggestion
    if( (PARSEC_FLOW_ACCESS_WRITE & type) && (gpu_task->task_type != PARSEC_GPU_TASK_TYPE_PREFETCH) )
```

devreal commented on parsec/mca/device/device_gpu.c:
```diff
@@ -1700,7 +1717,7 @@ parsec_device_callback_complete_push(parsec_device_gpu_module_t   *gpu_device,
                      * we do the command directly */
                     parsec_atomic_lock( &source->original->lock );
-                    int readers = parsec_atomic_fetch_sub_int32(&source->readers, 1);
+                    int readers = parsec_atomic_fetch_sub_int32(&source->readers, 1) - 1;
```
Curious: does that mean we were leaking items here because we never pushed them back to the LRU?

therault replied:
In the (somewhat rare) case when there was no GPU-manager for the source GPU, yes, we would leak some memory allocated via the `zone_malloc()`. The checks at termination would then complain, which is how we found the leak, and why we added the `zone_debug()` call at the completion checks in debug mode.

bosilca commented:
I can understand how the need to add debugging messages can be justified as part of this PR, but there are other things (such as changing the coherency state or moving data copy version manipulation code across function calls) that do not fit into the description of this PR.

Also, there are several instances where the indentation of the new code is incorrect, 8 commits (half of them not signed)  for a seemingly minor issue.

devreal commented on parsec/mca/device/device_gpu.c:
```diff
@@ -1339,6 +1331,10 @@ parsec_device_data_stage_in( parsec_device_gpu_module_t* gpu_device,
 
+        if( (PARSEC_FLOW_ACCESS_WRITE & type) && (gpu_task->task_type != PARSEC_GPU_TASK_TYPE_PREFETCH) ) {
+            gpu_elem->version = candidate->version + 1;
+		}
```
indentation

devreal commented on parsec/scheduling.c:
```diff
@@ -175,7 +175,9 @@ int __parsec_execute( parsec_execution_stream_t* es,
-        assert( NULL != hook );
+        if( NULL == hook ) {
+			goto next_chore; /* Gracefuly manage the case when NO chores are available for this task */
+		}
```
indentation

abouteiller replied:
I had made that same change on a prior pr and got shot down for it. Let me review this why it wasn't acceptable last time around.

abouteiller replied:
Found back the discussion https://github.com/ICLDisco/parsec/pull/616#discussion_r1462462587, I could not reproduce a scenario where that would happen, and added the assert so we would find that error easily when it shows up again, which is now.

devreal commented on tests/runtime/cuda/stage_main.c:
```diff
@@ -23,6 +23,7 @@ int main(int argc, char *argv[])
     int NB;
     int P = 1;
     int ret = 0;
+	int nb_gpus = 0;
```
indentation

devreal commented on tests/runtime/cuda/stage_main.c:
```diff
@@ -41,6 +42,19 @@ int main(int argc, char *argv[])
         exit(-1);
     }
 
+	for(int id = 0; id < (int)parsec_nb_devices; id++) {
+		parsec_device_module_t *device = parsec_mca_device_get(id);
+		if( PARSEC_DEV_IS_GPU(device->type) ) {
+			nb_gpus++;
+		}
+	}
+	if(nb_gpus == 0) {
+		fprintf(stderr, "This test can only run if at least one GPU device is present\n");
+		parsec_fini(&parsec);
+		MPI_Finalize();
+		return EXIT_FAILURE;
+	}
+
```
indentation

abouteiller commented:
> I can understand how the need to add debugging messages can be justified as part of this PR, but there are other things (such as changing the coherency state or moving data copy version manipulation code across function calls) that do not fit into the description of this PR.
> 
> Also, there are several instances where the indentation of the new code is incorrect, 8 commits (half of them not signed) for a seemingly minor issue.

The issue is not minor at all. Turning back on D2D management unearthed a whole bunch of bugs that would produce wrong results (in particular in TRSM where we would exercise CPU->GPU1,2,3->CPU data motions). While I agree that the description of the PR does not match the real scope of what is achieved here, the changes are not random and serve the greater case of making D2D work at a basic level.

abouteiller commented:
I will be extracting the fix for #641 in a separate PR, @therault that means I will be rewriting history in your branch to achieve this sorry.
